bpf/bpf_xdp.c

```diff
@@ -82,13 +81,35 @@ struct bpf_elf_map __section_maps CIDR6_LMAP_NAME = {
 #endif /* CIDR6_LPM_PREFILTER */
 #endif /* CIDR6_FILTER */
 
-static __always_inline int check_v4_endpoint(struct iphdr *ipv4_hdr)
+#ifdef ENABLE_IPV4
+#ifdef ENABLE_NODEPORT_ACCELERATION
+__section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV4_FROM_LXC)
+int tail_lb_ipv4(struct __ctx_buff *ctx)
 {
-	if (lookup_ip4_endpoint(ipv4_hdr))
-		return CTX_ACT_OK;
+	int ret = CTX_ACT_OK;
+
+	if (!bpf_skip_nodeport(ctx)) {
+		ret = nodeport_lb4(ctx, 0);
+		if (IS_ERR(ret))
+			return send_drop_notify_error(ctx, 0, ret, CTX_ACT_DROP,
+						      METRIC_INGRESS);
+	}
 
-	return CTX_ACT_DROP;
+	return ret;
+}
+
+static __always_inline int check_v4_lb(struct __ctx_buff *ctx)
+{
+	ep_tail_call(ctx, CILIUM_CALL_IPV4_FROM_LXC);
```

Comment:
Do we need the tail calls here? (even if we don't, we can always follow up after this PR..)

Reply:
From inside the nodeport handling we basically recirculate back to here (CILIUM_CALL_IPV4_FROM_LXC). Avoiding the tail call would be nice, but means we somehow need to get rid of the tail call chain inside our nodeport code. I can check if it's feasible in a follow-up.